File: objects.py
import pygame
from abc import ABC, abstractmethod
import math

pygame.init()

class Object: 
    def __init__(self, coords, color): 
        # unpack and assign coordinates
        x, y = coords
        self.x = x
        self.y = y
        self.color = color
    
    def set_coords(self, new_coords): 
        x, y = new_coords
        self.x = x
        self.y = y
    
    def get_coords(self): 
        return (self.x, self.y)

    @abstractmethod
    def draw(self, surface): 
        pass

class Circle(Object): 
    def __init__(self, radius, coords, color):
        super().__init__(coords, color)
        self.radius = radius
    
    def draw(self, surface): 
        pygame.draw.circle(surface, self.color, (self.x, self.y), self.radius)

class Square(Object): 
    def __init__(self, side, coords, color): 
        super().__init__(coords, color)
        self.side = side
    
    def draw(self, surface): 
        pygame.draw.rect(surface, self.color, (self.x, self.y, self.side, self.side))

class Polygon(Object): 
    # circumradius = distance from center to vertices
    def __init__(self, num_sides, circumradius, center_coords, color): 
        super().__init__(center_coords, color)
        vertices = calc_vertices(center_coords, num_sides, circumradius)
        self.vertices = vertices
        self.num_sides = num_sides
        self.circumradius = circumradius
    
    def set_coords(self, new_coords): 
        x, y = new_coords
        self.x = x
        self.y = y
        self.vertices = calc_vertices(new_coords, self.num_sides, self.circumradius)
    
    def get_vertices(self): 
        return self.vertices
    
    def draw(self, surface): 
        pygame.draw.polygon(surface, self.color, self.vertices)

def calc_vertices(center_coords, num_sides, circumradius): 
    # for the draw function later, and just to have
    vertices = []
    curr_angle = 90
    for side in range(num_sides): 
        vertices.append(calc_vertex_coord(center_coords, curr_angle, circumradius, side, num_sides))
        curr_angle += 360/num_sides
    return vertices

def calc_vertex_coord(cent_coords, curr_angle, circrad, vertex_num, num_vertices): 
    x, y = cent_coords
    cos_comp = circrad * math.cos((2*math.pi*vertex_num)/num_vertices)
    sin_comp = circrad * math.sin((2*math.pi*vertex_num)/num_vertices)
    x += cos_comp
    y += sin_comp
    return (x, y)

File: objects.js
class Object {
    constructor(coords, color) {
        // unpack and assign coordinates
        const [x, y] = coords;
        this.x = x;
        this.y = y;
        this.color = color;
    }

    setCoords(new_coords) {
        const [x, y] = new_coords;
        this.x = x;
        this.y = y;
    }

    getCoords() {
        return [this.x, this.y];
    }
}

class Circle extends Object {
    constructor(radius, coords, color) {
        super(coords, color);
        this.radius = radius;
    }

    draw(ctx) {
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.radius, 0, 2 * Math.PI);
        ctx.fillStyle = this.color;
        ctx.fill();
        ctx.stroke();
    }
}

class Square extends Object {
    constructor(side, coords, color) {
        super(coords, color);
        this.side = side;
    }
    
    draw(ctx) {
        ctx.fillStyle = this.color; 
        ctx.fillRect(this.x, this.y, this.side, this.side); 
    }
}

class Polygon extends Object {
    // circumradius = distance from center to vertices
    constructor(num_sides, circumradius, center_coords, color) {
        super(center_coords, color);
        this.vertices = calcVertices(center_coords, num_sides, circumradius);
        this.num_sides = num_sides;
        this.circumradius = circumradius;
    }
    setCoords(new_coords) {
        const [x, y] = new_coords;
        this.x = x;
        this.y = y;
        this.vertices = calcVertices(new_coords, this.num_sides, this.circumradius);
    }
    getVertices() {
        return this.vertices;
    }
    
    draw(ctx) {
        ctx.beginPath();
        ctx.moveTo(this.vertices[0][0], this.vertices[0][1]);
        for (let i = 1; i < this.vertices.length; i++) {
            ctx.lineTo(this.vertices[i][0], this.vertices[i][1]);
        }
        ctx.closePath();
        ctx.fillStyle = this.color;
        ctx.fill();
        ctx.stroke();
    }
}


function calcVertices(center_coords, num_sides, circumradius) {
    // for the draw function later, and just to have
    let vertices = [];
    let curr_angle = 90;
    for (let side = 0; side < num_sides; side++) {
        vertices.push(calcVertexCoord(center_coords, curr_angle, circumradius, side, num_sides));
        curr_angle += 360 / num_sides;
    }
    return vertices;
}


function calcVertexCoord(cent_coords, curr_angle, circrad, vertex_num, num_vertices) {
    let [x, y] = cent_coords;
    // calculate angle degree
    // converted to radians
    let angle_rad = curr_angle * Math.PI / 180;
    // update x and y based on sin and cosine
    x += circrad * Math.cos(angle_rad);
    y -= circrad * Math.sin(angle_rad);
    return [x, y];
}